### PR TITLE
Migrate from ObjectModel to Doctrine (final)

### PR DESCRIPTION
**Pull request by leemyongpakvn**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Please check details in the related issue.
| Type?         | refacto
| BC breaks?    | yes
| Deprecations? | yes
| Fixed ticket? | Fixes PrestaShop/Prestashop#33055
| How to test?  | Apply PR changes; Clear web browser cache & BO > Advanced Params > Performance > Clear Cache.<br> Check to make sure module works as usual in BO and FO.

### CODE REVIEW

File: src/Repository/ProductCommentRepository.php
Comment (matks) [inline src/Repository/ProductCommentRepository.php:239]: Please avoid [single letter variable](https://www.reddit.com/r/AskProgramming/comments/juebpx/what_are_your_thoughts_on_single_letter_variable/?rdt=50593)

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:66]: Because PHP test complained about it. In fact doctrine param type is declared right above.

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:130]: Sure, `ProductCommentCriterion::CATEGORIES_TYPE` is not a magic. 
I just want to keep old code as much as possible when migrating from ObjectModel ;)

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php:559]: After https://github.com/PrestaShop/ps_emailalerts/pull/129, I'm trying to not break the old logic, type and name.

**Comment by leemyongpakvn** [comment]
@matks Rebased and applied your suggestion.

File: src/Repository/ProductCommentCriterionRepository.php
Comment (kpodemski) [inline src/Repository/ProductCommentCriterionRepository.php:312]:
```suggestion
			SELECT pccc.id_category, pccc.id_product_comment_criterion
```
there's something wrong with indentation, I'm surprised cs fixer didn't catch it, you should indent using spaces, also the indentation itself is wrong for the queries

File: src/Repository/ProductCommentCriterionRepository.php
Comment (kpodemski) [inline src/Repository/ProductCommentCriterionRepository.php:333]: You could inject translator service in the construtor

File: src/Repository/ProductCommentRepository.php
Comment (kpodemski) [inline src/Repository/ProductCommentRepository.php:48]: I'm not a fun of such comments, what's the purpose of them?

File: src/Repository/ProductCommentRepository.php
Comment (kpodemski) [inline src/Repository/ProductCommentRepository.php:103]: I think you could simply remove this code if it is not used.

```suggestion
```

File: src/Repository/ProductCommentRepository.php
Comment (kpodemski) [inline src/Repository/ProductCommentRepository.php:281]:
```suggestion
```

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php]: Indentation and tabs are copied from [existing ObjectModel code](https://github.com/PrestaShop/productcomments/blob/v5.0.3/ProductCommentCriterion.php#L245-L248). I have just replaced tabs by spaces. Be free to change queries' indentation if you think it is needed.

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:333]: I tried inserting `@translator` in common.yml but got this error
> Fatal error: Uncaught Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "product_comment_criterion_repository" has a dependency on a non-existent service "translator".

Then I discovered that ps_linklist uses `@translator` in [its services](https://github.com/PrestaShop/ps_linklist/blob/dev/config/services.yml), `Symfony\Contracts\Translation\TranslatorInterface` in [its repository](https://github.com/PrestaShop/ps_linklist/blob/dev/src/Repository/LinkBlockRepository.php#L31), and dropped support for both PS 1.7.8, PS 8.0 :cry:

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php:42]: They are generated automatically when you run `Symfony make:entity` in terminal . I think it is useful in case someone try to reinvent the wheel ;)

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php]: It is a plan-for-future code. I don't care much of storing it explicitly in Release code, or storing it implicitly in github history.

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php]: as above ;)

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php]: We know php-cs-fixer and php-stan are not reliable in some cases.

File: src/Repository/ProductCommentRepository.php
Comment (matks) [reply src/Repository/ProductCommentRepository.php]: Do you know the concept of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) 😛 ?

What happens quite often with plan-for-future code is that we _think_ we're going to use it... but then we do something else 😅 and a few months later, another developer comes, he opens the code and he is confused by the code he sees. He does not know what to do with this commented code: is it important? should he follow it? is it something from the past?

This is why a lot of developers avoid commented out code https://methodpoet.com/commented-out-code/#:~:text=What%20is%20the%20commented%2Dout,issues%20if%20it's%20not%20removed.

Can I suggest to rather store this into a commit of yours, in git history, but do not put it in the release code? 😉

File: src/Repository/ProductCommentRepository.php
Comment (matks) [inline src/Repository/ProductCommentRepository.php:550]: There was no other hooks? I'm surprised only this one is in the PR body

File: src/Repository/ProductCommentCriterionRepository.php
Comment (matks) [inline src/Repository/ProductCommentCriterionRepository.php:209]: I maybe say something wrong but method name is `updateProducts()` however it does not update products. It creates new product_comment_criterion_product which indicate "this product comment is linked with this product".

So I see the creation but what about the deletion?

If I have comment A linked to products X, Y and Z.
I want to update comment A to link it to products X, Y and W. I dont want comment A to be linked to Z. This function cannot do it?

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php:550]: I see no other. It is migrated from [ProductComment.php](https://github.com/PrestaShop/productcomments/blob/dev/ProductComment.php#L354). Maybe @jolelievre knows its history because he saw it since [Version 4.0.0 version](https://github.com/PrestaShop/productcomments/pull/26)

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:209]: It is related to [src/Entity/ProductCommentCriterion.php](https://github.com/PrestaShop/productcomments/pull/177/files#diff-e62c70a054534c7be80a40ecc106507cd61896d8ed136cacc6b7c21abdab439cR86). In that context, `updateProducts()`  means update `products` that linked to  `$criterion`.
For deletion, you can see function [deleteProducts](https://github.com/PrestaShop/productcomments/pull/177/files#diff-1d36351690ee35c0a5e1d5c5c323743183c86e0239242510597a76805a72bebbR105-R110) at 100 lines above, in the same file.
If you check [the diff](https://github.com/PrestaShop/productcomments/pull/177/files#diff-1d36351690ee35c0a5e1d5c5c323743183c86e0239242510597a76805a72bebbR105-R110) you will see I don't change the logic: delete all old linked products, then add new linked products.

Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:209]: FYI, `updateProducts` and `deleteProducts` in  `ProductCommentCriterionRepository` is just a temporary solution when standard Doctrine `products` relation, `getProducts`, `addProduct`, `removeProduct` functions  are not implemented yet in `ProductCommentCriterion` Entity.

File: src/Repository/ProductCommentRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentRepository.php]: Yep, the plan-for-future code was wiped out from the Release code.

File: src/Repository/ProductCommentCriterionRepository.php
Comment (M0rgan01) [reply src/Repository/ProductCommentCriterionRepository.php:333]: With the upgrade to symfony 4, `Symfony\Component\Translation\TranslatorInterface` is transferred to `Symfony\Contracts\Translation\TranslatorInterface`. On the other hand, the translator service is still present, it has just changed the package.

**Review by matks — COMMENTED** [review]
Hi !

I have 1 good piece of news and 1 bad 😞 

The good news: for me this PR is valid 👍 I can approve it

The bad: this PR is modifying ProductComments module Doctrine entities, right?

Last week we released GDPR 2.0.1 module. This release also also modify module Doctrine entities. [And it started to _create failures for people trying to upgrade PrestaShop 8.x_.](https://github.com/PrestaShop/PrestaShop/discussions/32348#discussioncomment-6759336) So the module was rollbacked.

Right now, the problem is under analyzis but the main culprit is that _the way modules handle Doctrine schema updates migth break the shop upgrade process_.

So ... if we put this PR #177 inside release 6.0.0 of Productcomments, the same thing might happen. It might start crashing people shop upgrades 😞 .

So I'm really really really sad 😭 to say this but I'm afraid we should wait for the resolution of this upgrade problem before merging this PR. Because of the Doctrine entities being changed.

**Comment by kpodemski** [comment]
@matks when I check the code inside the class which is inside Entity directory, I don't see anything that could trigger schema upgrade

File: src/Repository/ProductCommentCriterionRepository.php
Comment (leemyongpakvn) [reply src/Repository/ProductCommentCriterionRepository.php:333]: Injecting translator service create error so I'd prefer do it in a minor version.

**Comment by leemyongpakvn** [comment]
I see 2 good news from **matks**'s bad news :satisfied:
First, the community can avoid a possible crash that is still under analyzed.
Second, I can bump the first **version 7** in native modules next month.

**Comment by matks** [comment]
Wait I might be wrong 😄 

https://github.com/PrestaShop/productcomments/blob/c85529e874056c82f68ab7daae0104ea3d26db5d/src/Entity/ProductCommentCriterion.php

@leemyongpakvn you added `private` properties in the Entity BUT you did not map them to SQL column so the Doctrine entities are NOT modified, right?

**Comment by leemyongpakvn** [comment]
@matks Correct, the ProductCommentCriterion Entity is MODIFIED, but 3 additional `private` properties is NOT mapped to SQL columns so as @kpodemski said
> I don't see anything that could trigger schema upgrade

File: src/Entity/ProductCommentCriterion.php
Comment (PululuK) [inline src/Entity/ProductCommentCriterion.php:91]:
```suggestion
```

File: src/Entity/ProductCommentCriterion.php
Comment (PululuK) [inline src/Entity/ProductCommentCriterion.php:216]:
```suggestion
    public function isValid()
    {
        foreach ($this->getNames() as $value) {
            if(!Validate::isGenericName($value)) {
                return false;
            }
        }

        return true;
    }
```

**Review by PululuK — COMMENTED** [review]
Small feedback :)

File: src/Entity/ProductCommentCriterion.php
Comment (PululuK) [reply src/Entity/ProductCommentCriterion.php:91]: Remove unused variable :)

File: src/Entity/ProductCommentCriterion.php
Comment (PululuK) [reply src/Entity/ProductCommentCriterion.php:216]: Performances improvement :)

**Comment by leemyongpakvn** [comment]
@matks In need of re-approve after removing unused variables (small tweak, no logic change ;)

File: src/Entity/ProductCommentCriterion.php
Comment (PululuK) [reply src/Entity/ProductCommentCriterion.php]: @leemyongpakvn You don't need to loop all the names... just have a false for the condition to return false

Comment (leemyongpakvn) [reply src/Entity/ProductCommentCriterion.php]: @PululuK You're right. Applied :)

**Comment by kpodemski** [comment]
@hibatallahAouadni modules don't require two approvals, it's ok 👍🏻

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @leemyongpakvn 

Thanks for your PR 🚀 
All module's functionalities works like expected ✅ 
@matks about the upgrade pb, do you want me test the upgrade with this PR? to check if there will be problems? 
Waiting for your feedback.

Thanks!

**Review by PululuK — APPROVED** [review]

**Review by hibatallahAouadni — APPROVED** [review]
As discussed with @kpodemski no need to test the upgrade 😉

**Pull request merged**
